Allow extenders to add content to order details screen payment info box

haszari opened the pull request:
This PR adds:

- A new hook to the order edit screen, so extenders can render info in the context of the payment details box. `woocommerce_admin_order_data_after_payment_info`
- ~~A new contextual notice component (PHP template + styles) that extenders can use to show a standard notice in context. `WC_Admin_Notices::render_contextual_notice()`~~ - removed in favour of Gutenberg / React `InlineNotice`

But why? 

In WooCommerce Payments we want to surface disputed orders clearly on the order edit screen – this is really important info for the merchant to action. 

- https://github.com/Automattic/woocommerce-payments/issues/6354

See companion PR https://github.com/Automattic/woocommerce-payments/pull/6373 where I use the new hook highlight disputes by displaying an inline notice (proposed component).


### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes # .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

##### Scenario 1: add a contextual notice in order details using React notice component
See https://github.com/Automattic/woocommerce-payments/pull/6373 for an example using new hook. 

1. Clone this branch & ensure you have at least one order.
2. In a custom plugin, implement the `woocommerce_admin_order_data_after_payment_info` hook.
3. In your hook function, render a container div.
4. Enqueue JavaScript that renders a React / Gutenberg / WC Admin notice in the container.

##### Scenario 2: display other arbitrary content in order payment details box

- As above; instead React notice, render some custom HTML, based on the current order.
- Is this a safe and useful hook to expose? 
- Should we lock it down more?

---

#### Old / no longer relevant test instructions below

(Old testing instructions hidden in `details` tag below)
<details>
##### Scenario 1: add a contextual notice in order details

See https://github.com/Automattic/woocommerce-payments/pull/6373 for an example using new hook. Details for how to override

1. Clone this branch & ensure you have at least one order.
2. In a custom plugin, implement the `woocommerce_admin_order_data_after_payment_info` hook.
3. Try different values for `$severity`, `$button_text` and $button_url`.
  - If either button option is missing or empty, the button should not render.
  - Severity renders a notice with different colours: 
    - `error`=red.
    - `warning`=yellow.
    - `success`=green.
    - `info`=blue (default).

Code example:

```php
// Add this to a custom plugin.


add_action( 'woocommerce_admin_order_data_after_payment_info', [ $this, 'myplugin_test_contextual_notice' ] );

public function myplugin_test_contextual_notice() {
	if ( ! method_exists( 'WC_Admin_Notices', 'render_contextual_notice' ) ) {
		return;
	}

	$notice_html = "This order has an active dispute for $25. Please respond today to avoid penalties.";
	$button_text = "Respond now";
	$severity = 'error';
	$button_url = 'https://woocommerce.com'; // Link to dispute details!
	WC_Admin_Notices::render_contextual_notice( $notice_html, $severity, $button_text, $button_url );

	$notice_html = "This order has an active dispute for $25. Please respond within 7 days.";
	$severity = 'warning';
	$button_text = "Respond now";
	$button_url = 'https://woocommerce.com'; // Link to dispute details!
	WC_Admin_Notices::render_contextual_notice( $notice_html, $severity, $button_text, $button_url );

	$notice_html = "Payment for this order has not been captured. Please capture payment within 7 days.";
	$severity = 'info';
	$button_text = "Capture payment";
	$button_url = 'https://woocommerce.com'; // Link to dispute details!
	WC_Admin_Notices::render_contextual_notice( $notice_html, $severity, $button_text, $button_url );

	$notice_html = "This order is shipped and completed.";
	$severity = 'success';
  WC_Admin_Notices::render_contextual_notice( $notice_html, $severity );
}
```

![Screenshot 2023-05-26 at 11 26 21 AM](https://github.com/woocommerce/woocommerce/assets/4167300/92673eef-c6b4-4f85-b95d-1c85a20dc92e)
</details>

(`render_contextual_notice` test instructions no longer relevant, old details in `details`)
<details>
##### Scenario 3: use `WC_Admin_Notices::render_contextual_notice` elsewhere in UI

- Use `WC_Admin_Notices::render_contextual_notice()` to render similar contextual notices in the UI.
- Does the notice render correctly in other contexts?
- Can you think of other use cases that we should consider?
- Is this useful in WC core?
</details>

Commit:
add a hook to allow extenders to add notices or content after payment details on order edit screen

Commit:
Add fake but informative changelog to appease the pre-commit hooks.

Commit:
prototype contextual notice component – php func, template, and css

coreymckrill commented on plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-order-data.php:
```diff
@@ -250,6 +250,9 @@ public static function output( $post ) {
 
 					?>
 				</p>
+				<div>
+					<?php do_action( 'woocommerce_order_details_after_payment_info' ); ?>
+				</div>
```
A few thoughts here:

* I would recommend just firing the action without adding an extra wrapper to it (since the wrapper would be empty most of the time). The callback can add any wrapper necessary.
* For naming consistency with other action hooks in this meta box, such as `woocommerce_admin_order_data_after_order_details`, what about something like this for the hook name: `woocommerce_admin_order_data_after_payment_info`
* Probably worth adding `$order` as a parameter, like the other actions do.

haszari replied:
Sounds good, I've made these changes and added a docblock: ff89bfda40a2d0928054e96869b71a8838cb6453

> I would recommend just firing the action without adding an extra wrapper to it (since the wrapper would be empty most of the time). The callback can add any wrapper necessary.

My thinking here was to add a wrapper div (with a class?) to limit the damage that extenders can do - a sandbox of sorts. I.e. make the hook more tailored to specific supported use cases (taken to extreme, could be a `_contextual_notice` filter). Anyway I've removed the div.

coreymckrill commented on plugins/woocommerce/includes/admin/views/html-notice-contextual.php:
```diff
@@ -0,0 +1,19 @@
+?>
+<div class="woocommerce-contextual-notice woocommerce-contextual-notice-<?php echo esc_attr( $severity ); ?>">
+  <?php echo wp_kses_post( $notice_html ); ?>
+  <div class="woocommerce-contextual-notice-cta" >
+  	<a class="components-button is-secondary"
+      href="<?php echo esc_url( $button_url ); ?>">
+      <?php echo esc_html( $button_text ) ?>
+    </a>
+  </div>
```
I'd probably make the button parameters optional, and only conditionally render this part.

haszari replied:
I agree, hadn't got to that yet! Now all params except the body HTML are optional, and the button will only render if both params are provided. bb79e28cf6a7884c23613017c339d5e71d2a2a4c

We could also add a "dismissable" option. I'm hesitant to add it as I don't need it for my use case, and maybe these contextual notices should be reserved for things that shouldn't be dismissable. Can bring in design folks to refine how the new component should work & what it should allow.

coreymckrill commented on plugins/woocommerce/includes/admin/class-wc-admin-notices.php:
```diff
@@ -252,6 +252,24 @@ public static function add_custom_notice( $name, $notice_html ) {
+	 * @param string $button_text Button text (plain text).
+	 * @param string $button_url  URL to link to.
+	 */
+	public static function render_contextual_notice( $notice_html, $button_text, $severity, $button_url ) {
```
It looks like `$severity` is supposed to come before `$button_text` in the parameter order?

haszari replied:
Good spotting! Fixed here 4c9df162ce428997f54d3356c49ab991d0329b52

Commit:
fix parameter order – consistent with docblock, group soon-tobe-optionall button params

Commit:
tidy / polish new order metabox hook:
- rename for consistency with related hooks
- add `$order` param as will likely be useful to clients
- document in docblock
- remove wrapper div

Commit:
make all params optional (except body text) and conditonally render button

haszari commented on plugins/woocommerce/client/legacy/css/admin.scss:
```diff
@@ -8019,3 +8019,52 @@ table.bar_chart {
+		box-shadow: inset 0 0 0 1px #CC1818;
+		border-color: #CC1818;
+		color: #CC1818;
+	}
```
This CSS is all working for my purposes, but I'm keen to refine it a lot more, with guidance from experts & testing from reviewers. Also design folks could have useful input here.

1. The colours used are all official / Woo / Gutenberg colours. Do we have scss variables I can use instead of pasting raw hex?
2. I've overridden existing `.components-button` (to change colours). Need to stress test this and ensure there aren't any hover/active etc states which are showing wrong (default) colours.
3. Class naming – I'm keen for this to be a reusable component, and to name the classes consistent with however we do that. e.g. `.component-contextual-notice` etc.
4. Component naming - I've named this `woocommerce-contextual-notice`; keen for ideas for the right name for this. Could be semantic – name for specific purpose (i.e. very important, non-dismissable notices in context); or low-level – name as raw elements that can be used for any purpose.
5. Padding / margins – I've hacked to get the layout I need. Might need some more thought, care, testing to get this right for general use (e.g. in other contexts).

Jinksi replied:
I notice that the disputes project Figma design (RiEQmaKRI7u9USAI3lyZz0VX-fi-2885_194324) uses the component "Inline Notice" from the WC Admin design library: Kofpo8jj4zDWXDsVdlSz0f-fi-4254_46957.

Based on this, should we consider the naming `woocommerce-inline-notice`?

Does this component and associated styles already exist within the WC codebase?

haszari replied:
> Based on this, should we consider the naming woocommerce-inline-notice?

> Does this component and associated styles already exist within the WC codebase?

This is similar to @louwie17  [feedback](https://github.com/woocommerce/woocommerce/pull/38443#pullrequestreview-1455343676) 😁 

> We also encourage the use of the React notice component from Gutenberg, and as more and more of the UI is being replaced by React, using the notice component would make more sense.

I'm going to rework to just add the hook and then use the Gutenberg / React notice component to render a nice notice inline.

haszari commented:
Reviewer notes / questions:

- I have not added an issue / feature request for this. Happy to do so, let me know.
- The naming and behaviour of the `contextual notice` component is very much up for discussion. Keen for feedback from transact design @elizaan36  and WC core design also.
- There are two new APIs added in this PR. These are both up for discussion / feedback also.
  - `woocommerce_admin_order_data_after_payment_info` - generic hook allowing any content to be injected into order details box after payment line
  - `WC_Admin_Notices::render_contextual_notice()` - render a contextual notice.
  - Specific API questions:
    - Are each of these APIs useful and necessary and safe; or should we lock this down more?
    - Alternative: should we expose a much more limited `woocommerce_admin_order_data_after_payment_info_contextual_notice` filter, so extenders can't add arbitrary content?
- If we proceed with "contextual notice", the CSS needs more testing & review.

haszari commented:
I've requested reviews from Helix (couldn't tag team, all welcome but only require one team review).

Also tagged for design feedback.

Thanks for the initial review @coreymckrill  – feel free to have another look/test. I'm keen for a wcadmin / wccore front end expert to review (CSS /component). Keen for thoughts on the new APIs in general.

haszari commented:
Tagged @woocommerce/mothra  for a review: 

- PR proposes a new merchant-admin facing hook: `woocommerce_admin_order_data_after_payment_info`.
- PR proposes a new visual / UI component for contextual (or inline) notices, and an API for rendering such notices (in PHP; could also implement a JS / front end API): `WC_Admin_Notices::render_contextual_notice()`.

louwie17 commented on plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-order-data.php:
```diff
@@ -250,6 +250,24 @@ public static function output( $post ) {
+					 * used to render a prominent notice with a CTA button.
+					 *
+					 * @param $order WC_Order The order object being displayed.
+					 * @since 3.0.0
```
This would be 7.8.0 if we do decide to merge this.

louwie17 reviewed (changes requested):
Overall this looks good, but I would push back against the addition of the contextual notices and just stick with the new hook within the order meta box.
At the moment I don't see to many additional use cases for adding the contextual notice in other places, where we don't already support notices. We also encourage the use of the React notice component from Gutenberg, and as more and more of the UI is being replaced by React, using the notice component would make more sense.

Lastly, I think we should be wary about introducing notices in a lot of places anyhow, as from my experience users tend to dislike them and find them very disruptive.

Just adding the hook will unblock your work within WooCommerce Payments, and gives you more freedom to implement it how you see fit.

haszari commented:
> Overall this looks good, but I would push back against the addition of the contextual notices and just stick with the new hook within the order meta box.

Thanks @louwie17 – I can work with that. 

I'll revise my approach to implement the note markup in WCPay. We can also look at using Gutenberg components or sharing styles rather than having a parallel PHP implementation of essentially the same component.

Commit:
remove `contextual notice` experiment, can use inline notice Gutenberg component

Commit:
correct @since version and clarify changelog

haszari commented:
I've updated this PR to just expose the new hook `woocommerce_admin_order_data_after_payment_info`.

- I'm targeting WC 7.9 for this (at the earliest).

I'm updating [companion PR in WCPay](https://github.com/Automattic/woocommerce-payments/pull/6373) to use Gutenberg / WC component `InlineNotice` to render the notice (instead of custom / redundant `contextual notice`, now removed). After I've got that working I'll re-request reviews.

louwie17 reviewed (approved):
Thanks for the update @haszari, this looks good from my end 🚀 ✅ 
Feel free to merge when you can.

haszari commented:
Thanks for the feedback & assistance getting this shipshape @louwie17 @coreymckrill !

> Feel free to merge when you can.

I can't merge - I don't have permission. Can someone merge on my behalf? Or grant me permissions if that's better. Thanks 😁 

<img width="897" alt="Screenshot 2023-06-09 at 10 20 03 AM" src="https://github.com/woocommerce/woocommerce/assets/4167300/fc03392b-6ce2-40b8-af01-4f6f24df9b3f">

coreymckrill commented on plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-order-data.php:
```diff
@@ -250,6 +250,24 @@ public static function output( $post ) {
+					 * the order page. Example: display a notice if
+					 * the order is disputed.
+					 *
+					 * WC_Admin_Notices::render_contextual_notice() can be
+					 * used to render a prominent notice with a CTA button.
```
@haszari This method got removed, right?

haszari replied:
Thanks for catching this ! I'll remove and re-check to ensure there aren't any other remnants.

haszari replied:
Removed! e8f2d3f0618040dbe421177fe6f5d09d59fe39a9 

Should be good to merge @coreymckrill  – let me know if there's anything else.

Commit:
remove remnant comment about "contextual notice" (now gone):
- intending to use Gutenberg notice and/or proposed forthcoming InlineNotice variant
